{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "482bb4fa",
   "metadata": {},
   "source": [
    "# PhageHostPredict: data processing\n",
    "\n",
    "In this notebook, all the necessary data processing steps are completed in order to run PhageHostPredict models with RBPs and loci. The notebook transforms collections of phage genomes and bacterial genomes to collections of RBPs and loci."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b0c9db",
   "metadata": {},
   "source": [
    "## Libraries & files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "227a6436",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import json\n",
    "import math\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import processing_utils as pu\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from tqdm import tqdm\n",
    "from os import listdir\n",
    "from xgboost import XGBClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40bfe57",
   "metadata": {},
   "source": [
    "In the block below, set the correct directories and locations. Both the phage genomes and bacterial genomes directory should be folders with all genomes present in them. To run Phanotate, the location of the python file should also be provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d875d055",
   "metadata": {},
   "outputs": [],
   "source": [
    "valencia_dir = '/Users/dimi/GoogleDrive/PhD/4_PHAGEHOST_LEARNING/42_DATA/Valencia_data' # general directory\n",
    "phage_genomes_dir = valencia_dir+'/phages_genomes'\n",
    "bacterial_genomes_dir = valencia_dir+'/klebsiella_genomes/fasta_files'\n",
    "phanotate_loc = '/opt/homebrew/Caskroom/miniforge/base/envs/ML1/bin/phanotate.py'\n",
    "project_dir = '/Users/dimi/Documents/GitHub_Local/PhageHostLearning'\n",
    "data_suffix = 'Valencia' # choose a suffix for the created data files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8609966c",
   "metadata": {},
   "source": [
    "## Phage genome processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a56cf9b",
   "metadata": {},
   "source": [
    "#### Phanotate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c9d2d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "phage_files = listdir(phage_genomes_dir)\n",
    "phage_files.remove('.DS_Store')\n",
    "#record = SeqIO.read(phages_dir+'/'+phage_files[0], 'fasta')\n",
    "bar = tqdm(total=len(phage_files), position=0, leave=True)\n",
    "name_list = []; gene_list = []\n",
    "\n",
    "for file in phage_files:\n",
    "    # access PHANOTATE\n",
    "    file_dir = phage_genomes_dir+'/'+file\n",
    "    raw_str = phanotate_loc + ' ' + file_dir\n",
    "    process = subprocess.Popen(raw_str, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)\n",
    "    stdout, stderr = process.communicate()\n",
    "    std_splits = stdout.split(sep=b'\\n')\n",
    "    std_splits = std_splits[2:] #std_splits.pop(0)\n",
    "    \n",
    "    # Save and reload TSV\n",
    "    temp_tab = open('/Users/Dimi/Desktop/phage_results.tsv', 'wb')\n",
    "    for split in std_splits:\n",
    "        split = split.replace(b',', b'') # replace commas for pandas compatibility\n",
    "        temp_tab.write(split + b'\\n')\n",
    "    temp_tab.close()\n",
    "    results_orfs = pd.read_csv('/Users/Dimi/Desktop/phage_results.tsv', sep='\\t', lineterminator='\\n', index_col=False)\n",
    "    \n",
    "    # fill up lists accordingly\n",
    "    name = file.split('_')[0]\n",
    "    sequence = str(SeqIO.read(file_dir, 'fasta').seq)\n",
    "    for j, strand in enumerate(results_orfs['FRAME']):\n",
    "        start = results_orfs['#START'][j]\n",
    "        stop = results_orfs['STOP'][j]\n",
    "        \n",
    "        if strand == '+':\n",
    "            gene = sequence[start-1:stop]\n",
    "        else:\n",
    "            sequence_part = sequence[stop-1:start]\n",
    "            gene = str(Seq(sequence_part).reverse_complement())\n",
    "            \n",
    "        name_list.append(name)\n",
    "        gene_list.append(gene)\n",
    "        \n",
    "    # update progress\n",
    "    bar.update(1)\n",
    "bar.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f5c13b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export final genes database\n",
    "genebase = pd.DataFrame(list(zip(name_list, gene_list)), columns=['phage_ID', 'gene_sequence'])\n",
    "genebase.to_csv(valencia_dir+'/phage_genes'+data_suffix+'.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "439329de",
   "metadata": {},
   "source": [
    "#### RBP detection\n",
    "\n",
    "In the code block below, provide the correct paths to the Pfam file, the trained XGB model, the saved phage genes, and protein embeddings of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "45c7c89d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam_file = '/Users/dimi/Documents/GitHub/PhageHostLearning/processing/RBPdetect_phageRBPs.hmm'\n",
    "xgb_file = 'RBPdetect_xgb_model.json'\n",
    "#fasta_file = 'data/examples.fasta'\n",
    "hmmer_path = '/Users/Dimi/hmmer-3.3.1' # e.g. '/Users/Sally/hmmer-3.3.1'\n",
    "embeddings_file = ... # e.g. 'data/embeddings.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bbc07d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# optionally press database first if not done already\n",
    "#output, err = pu.hmmpress_python(hmmer_path, pfam_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "72a0c866",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scanning the genes: 100%|███████████████████| 5657/5657 [01:35<00:00, 59.33it/s]\n"
     ]
    }
   ],
   "source": [
    "# domain-based RBP detection\n",
    "N_blocks = ['Phage_T7_tail', 'Tail_spike_N', 'Prophage_tail', 'BppU_N', 'Mtd_N', \n",
    "           'Head_binding', 'DUF3751', 'End_N_terminal', 'phage_tail_N', 'Prophage_tailD1', \n",
    "           'DUF2163', 'Phage_fiber_2', 'unknown_N0', 'unknown_N1', 'unknown_N2', 'unknown_N3', 'unknown_N4', \n",
    "            'unknown_N6', 'unknown_N10', 'unknown_N11', 'unknown_N12', 'unknown_N13', 'unknown_N17', 'unknown_N19', \n",
    "            'unknown_N23', 'unknown_N24', 'unknown_N26','unknown_N29', 'unknown_N36', 'unknown_N45', 'unknown_N48', \n",
    "            'unknown_N49', 'unknown_N53', 'unknown_N57', 'unknown_N60', 'unknown_N61', 'unknown_N65', 'unknown_N73', \n",
    "            'unknown_N82', 'unknown_N83', 'unknown_N101', 'unknown_N114', 'unknown_N119', 'unknown_N122', \n",
    "            'unknown_N163', 'unknown_N174', 'unknown_N192', 'unknown_N200', 'unknown_N206', 'unknown_N208']\n",
    "C_blocks = ['Lipase_GDSL_2', 'Pectate_lyase_3', 'gp37_C', 'Beta_helix', 'Gp58', 'End_beta_propel', \n",
    "            'End_tail_spike', 'End_beta_barrel', 'PhageP22-tail', 'Phage_spike_2', \n",
    "            'gp12-short_mid', 'Collar', \n",
    "            'unknown_C2', 'unknown_C3', 'unknown_C8', 'unknown_C15', 'unknown_C35', 'unknown_C54', 'unknown_C76', \n",
    "            'unknown_C100', 'unknown_C105', 'unknown_C112', 'unknown_C123', 'unknown_C179', 'unknown_C201', \n",
    "            'unknown_C203', 'unknown_C228', 'unknown_C234', 'unknown_C242', 'unknown_C258', 'unknown_C262', \n",
    "            'unknown_C267', 'unknown_C268', 'unknown_C274', 'unknown_C286', 'unknown_C292', 'unknown_C294', \n",
    "            'Peptidase_S74', 'Phage_fiber_C', 'S_tail_recep_bd', 'CBM_4_9', 'DUF1983', 'DUF3672']\n",
    "\n",
    "phage_genes = pd.read_csv(valencia_dir+'/phage_genes'+data_suffix+'.csv')\n",
    "phage_genes_list = phage_genes['gene_sequence']\n",
    "phage_genes_ids = phage_genes['phage_ID']\n",
    "\n",
    "domain_based_detections = pu.RBPdetect_domains(hmmer_path, pfam_file, phage_genes_list, phage_genes_ids, \n",
    "                                               N_blocks=N_blocks, C_blocks=C_blocks, detect_others=False)\n",
    "\n",
    "domain_preds = []\n",
    "for pid in phage_genes_ids:\n",
    "    if pid in list(domain_based_detections['identifier']):\n",
    "        domain_preds.append(1)\n",
    "    else:\n",
    "        domain_preds.append(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "86728209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>DNASeq</th>\n",
       "      <th>N_blocks</th>\n",
       "      <th>C_blocks</th>\n",
       "      <th>N_ranges</th>\n",
       "      <th>C_ranges</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K39PH122C2</td>\n",
       "      <td>ATGGCATTCAGCTGGCAAGAGTCGGTAAAGCCTGCAGGTACTCAGG...</td>\n",
       "      <td>[Phage_T7_tail, unknown_N2]</td>\n",
       "      <td>[unknown_C76, unknown_C234, unknown_C286]</td>\n",
       "      <td>[(12, 140), (3, 105)]</td>\n",
       "      <td>[(140, 791), (142, 263), (26, 37)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K39PH122C2</td>\n",
       "      <td>ATGGCTTTAGTAAAACTTACACGGGTAGCGGAGTGGCTCGGAACGT...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Pectate_lyase_3, Peptidase_S74]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(38, 126), (543, 603)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K60PH164C1</td>\n",
       "      <td>ATGAATTCTCATAACCCGTTTAACACAGGTGGTAGTGGGTGCACCA...</td>\n",
       "      <td>[unknown_N82]</td>\n",
       "      <td>[unknown_C105]</td>\n",
       "      <td>[(124, 237)]</td>\n",
       "      <td>[(235, 331)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K45PH128C2</td>\n",
       "      <td>ATGGCTGATACAACACAATTTGAACAGGCTGTGGATCAGGTTATCG...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[unknown_C123]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(66, 115)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K45PH128C2</td>\n",
       "      <td>ATGGCAATTCCTACTATTCCTCTTCAGATATGGGCTGAAAGTGACG...</td>\n",
       "      <td>[unknown_N48]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 82)]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>K33PH14C2</td>\n",
       "      <td>ATGACTAATATCAAGGCCCGCAAGGGCGGTTCCAGCCAGCCGCGAA...</td>\n",
       "      <td>[unknown_N12, unknown_N163, unknown_N206, unkn...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 199), (3, 186), (2, 198), (3, 194), (9, 1...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>K33PH14C2</td>\n",
       "      <td>ATGGACGCAAACGGCGTTATCACCGGGACTGGTACGAAGTGGCAAT...</td>\n",
       "      <td>[unknown_N11]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 163)]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>K33PH14C2</td>\n",
       "      <td>CTGGATTCGTGGGCCAATGTTGATGGCGTAGGTTCCATGTACACAA...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[DUF1983]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(2, 77)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>K35PH164C3</td>\n",
       "      <td>ATGGATAATGAAATACGCACGGTGGTCACGTACAACCTAGATGGGG...</td>\n",
       "      <td>[Phage_T7_tail]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 117)]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>K43PH164C1</td>\n",
       "      <td>ATGGACCAAGAAATTAAAACAGTCATCCAGTACCCAGTAGGGGCCA...</td>\n",
       "      <td>[Phage_T7_tail]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 154)]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    identifier                                             DNASeq  \\\n",
       "0   K39PH122C2  ATGGCATTCAGCTGGCAAGAGTCGGTAAAGCCTGCAGGTACTCAGG...   \n",
       "1   K39PH122C2  ATGGCTTTAGTAAAACTTACACGGGTAGCGGAGTGGCTCGGAACGT...   \n",
       "2   K60PH164C1  ATGAATTCTCATAACCCGTTTAACACAGGTGGTAGTGGGTGCACCA...   \n",
       "3   K45PH128C2  ATGGCTGATACAACACAATTTGAACAGGCTGTGGATCAGGTTATCG...   \n",
       "4   K45PH128C2  ATGGCAATTCCTACTATTCCTCTTCAGATATGGGCTGAAAGTGACG...   \n",
       "..         ...                                                ...   \n",
       "89   K33PH14C2  ATGACTAATATCAAGGCCCGCAAGGGCGGTTCCAGCCAGCCGCGAA...   \n",
       "90   K33PH14C2  ATGGACGCAAACGGCGTTATCACCGGGACTGGTACGAAGTGGCAAT...   \n",
       "91   K33PH14C2  CTGGATTCGTGGGCCAATGTTGATGGCGTAGGTTCCATGTACACAA...   \n",
       "92  K35PH164C3  ATGGATAATGAAATACGCACGGTGGTCACGTACAACCTAGATGGGG...   \n",
       "93  K43PH164C1  ATGGACCAAGAAATTAAAACAGTCATCCAGTACCCAGTAGGGGCCA...   \n",
       "\n",
       "                                             N_blocks  \\\n",
       "0                         [Phage_T7_tail, unknown_N2]   \n",
       "1                                                  []   \n",
       "2                                       [unknown_N82]   \n",
       "3                                                  []   \n",
       "4                                       [unknown_N48]   \n",
       "..                                                ...   \n",
       "89  [unknown_N12, unknown_N163, unknown_N206, unkn...   \n",
       "90                                      [unknown_N11]   \n",
       "91                                                 []   \n",
       "92                                    [Phage_T7_tail]   \n",
       "93                                    [Phage_T7_tail]   \n",
       "\n",
       "                                     C_blocks  \\\n",
       "0   [unknown_C76, unknown_C234, unknown_C286]   \n",
       "1            [Pectate_lyase_3, Peptidase_S74]   \n",
       "2                              [unknown_C105]   \n",
       "3                              [unknown_C123]   \n",
       "4                                          []   \n",
       "..                                        ...   \n",
       "89                                         []   \n",
       "90                                         []   \n",
       "91                                  [DUF1983]   \n",
       "92                                         []   \n",
       "93                                         []   \n",
       "\n",
       "                                             N_ranges  \\\n",
       "0                               [(12, 140), (3, 105)]   \n",
       "1                                                  []   \n",
       "2                                        [(124, 237)]   \n",
       "3                                                  []   \n",
       "4                                           [(0, 82)]   \n",
       "..                                                ...   \n",
       "89  [(0, 199), (3, 186), (2, 198), (3, 194), (9, 1...   \n",
       "90                                         [(0, 163)]   \n",
       "91                                                 []   \n",
       "92                                         [(0, 117)]   \n",
       "93                                         [(0, 154)]   \n",
       "\n",
       "                              C_ranges  \n",
       "0   [(140, 791), (142, 263), (26, 37)]  \n",
       "1              [(38, 126), (543, 603)]  \n",
       "2                         [(235, 331)]  \n",
       "3                          [(66, 115)]  \n",
       "4                                   []  \n",
       "..                                 ...  \n",
       "89                                  []  \n",
       "90                                  []  \n",
       "91                           [(2, 77)]  \n",
       "92                                  []  \n",
       "93                                  []  \n",
       "\n",
       "[94 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domain_based_detections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c42c95e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ML-based RBP detection\n",
    "embeddings_df = pd.read_csv(embeddings_file)\n",
    "embeddings = np.asarray(embeddings_df.iloc[:, 2:])\n",
    "\n",
    "xgb_saved = XGBClassifier()\n",
    "xgb_saved.load_model(xgb_file) # load trained model\n",
    "\n",
    "score_xgb = xgb_saved.predict_proba(embeddings)[:,1]\n",
    "preds_xgb = (score_xgb > 0.5)*1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61c1d6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# take union or intersection of both detection methods\n",
    "..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "11795ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a dataframe and unique protein IDs\n",
    "proteins = [str(Seq(sequence).translate())[:-1] for sequence in domain_based_detections['DNASeq']]\n",
    "unique_ids = []\n",
    "for item in domain_based_detections['identifier']:\n",
    "    check = 0; count = 0\n",
    "    while check == 0:\n",
    "        unique = item+'_RBP'+str(count)\n",
    "        if unique not in unique_ids:\n",
    "            unique_ids.append(unique)\n",
    "            check = 1\n",
    "        else:\n",
    "            count += 1\n",
    "            \n",
    "RBPs = pd.DataFrame({'phage_nr': domain_based_detections['identifier'], \n",
    "                      'host': ['klebsiella_pneumoniae']*len(domain_based_detections['identifier']), \n",
    "                      'host_accession': ['-']*len(domain_based_detections['identifier']), \n",
    "                      'DNASeq': domain_based_detections['DNASeq'], \n",
    "                      'ProteinSeq': proteins, 'N_blocks': domain_based_detections['N_blocks'], \n",
    "                      'C_blocks': domain_based_detections['C_blocks'], 'N_ranges': domain_based_detections['N_ranges'], \n",
    "                      'C_ranges': domain_based_detections['C_ranges'], 'unique_ID': unique_ids})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9ef1f0ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phage_nr</th>\n",
       "      <th>host</th>\n",
       "      <th>host_accession</th>\n",
       "      <th>DNASeq</th>\n",
       "      <th>ProteinSeq</th>\n",
       "      <th>N_blocks</th>\n",
       "      <th>C_blocks</th>\n",
       "      <th>N_ranges</th>\n",
       "      <th>C_ranges</th>\n",
       "      <th>unique_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K39PH122C2</td>\n",
       "      <td>klebsiella_pneumoniae</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGCATTCAGCTGGCAAGAGTCGGTAAAGCCTGCAGGTACTCAGG...</td>\n",
       "      <td>MAFSWQESVKPAGTQDIQCDIEYLDKSYIHVYLDGAETTAFTWTSS...</td>\n",
       "      <td>[Phage_T7_tail, unknown_N2]</td>\n",
       "      <td>[unknown_C76, unknown_C234, unknown_C286]</td>\n",
       "      <td>[(12, 140), (3, 105)]</td>\n",
       "      <td>[(140, 791), (142, 263), (26, 37)]</td>\n",
       "      <td>K39PH122C2_RBP0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K39PH122C2</td>\n",
       "      <td>klebsiella_pneumoniae</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGCTTTAGTAAAACTTACACGGGTAGCGGAGTGGCTCGGAACGT...</td>\n",
       "      <td>MALVKLTRVAEWLGTYIHSLSSVQRLLGSKLDDTLCVLDFGAVADY...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Pectate_lyase_3, Peptidase_S74]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(38, 126), (543, 603)]</td>\n",
       "      <td>K39PH122C2_RBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K60PH164C1</td>\n",
       "      <td>klebsiella_pneumoniae</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGAATTCTCATAACCCGTTTAACACAGGTGGTAGTGGGTGCACCA...</td>\n",
       "      <td>MNSHNPFNTGGSGCTNDFRSADPLVDRIIGDAYHVVKEVYLALGNL...</td>\n",
       "      <td>[unknown_N82]</td>\n",
       "      <td>[unknown_C105]</td>\n",
       "      <td>[(124, 237)]</td>\n",
       "      <td>[(235, 331)]</td>\n",
       "      <td>K60PH164C1_RBP0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K45PH128C2</td>\n",
       "      <td>klebsiella_pneumoniae</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGCTGATACAACACAATTTGAACAGGCTGTGGATCAGGTTATCG...</td>\n",
       "      <td>MADTTQFEQAVDQVIEDSERLHKVVNGSAIDTVIVEDGSTIPTLRK...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[unknown_C123]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(66, 115)]</td>\n",
       "      <td>K45PH128C2_RBP0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K45PH128C2</td>\n",
       "      <td>klebsiella_pneumoniae</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGCAATTCCTACTATTCCTCTTCAGATATGGGCTGAAAGTGACG...</td>\n",
       "      <td>MAIPTIPLQIWAESDVVLPNAHTANKISPIADLWDKGWDLGEKPAC...</td>\n",
       "      <td>[unknown_N48]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[(0, 82)]</td>\n",
       "      <td>[]</td>\n",
       "      <td>K45PH128C2_RBP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     phage_nr                   host host_accession  \\\n",
       "0  K39PH122C2  klebsiella_pneumoniae              -   \n",
       "1  K39PH122C2  klebsiella_pneumoniae              -   \n",
       "2  K60PH164C1  klebsiella_pneumoniae              -   \n",
       "3  K45PH128C2  klebsiella_pneumoniae              -   \n",
       "4  K45PH128C2  klebsiella_pneumoniae              -   \n",
       "\n",
       "                                              DNASeq  \\\n",
       "0  ATGGCATTCAGCTGGCAAGAGTCGGTAAAGCCTGCAGGTACTCAGG...   \n",
       "1  ATGGCTTTAGTAAAACTTACACGGGTAGCGGAGTGGCTCGGAACGT...   \n",
       "2  ATGAATTCTCATAACCCGTTTAACACAGGTGGTAGTGGGTGCACCA...   \n",
       "3  ATGGCTGATACAACACAATTTGAACAGGCTGTGGATCAGGTTATCG...   \n",
       "4  ATGGCAATTCCTACTATTCCTCTTCAGATATGGGCTGAAAGTGACG...   \n",
       "\n",
       "                                          ProteinSeq  \\\n",
       "0  MAFSWQESVKPAGTQDIQCDIEYLDKSYIHVYLDGAETTAFTWTSS...   \n",
       "1  MALVKLTRVAEWLGTYIHSLSSVQRLLGSKLDDTLCVLDFGAVADY...   \n",
       "2  MNSHNPFNTGGSGCTNDFRSADPLVDRIIGDAYHVVKEVYLALGNL...   \n",
       "3  MADTTQFEQAVDQVIEDSERLHKVVNGSAIDTVIVEDGSTIPTLRK...   \n",
       "4  MAIPTIPLQIWAESDVVLPNAHTANKISPIADLWDKGWDLGEKPAC...   \n",
       "\n",
       "                      N_blocks                                   C_blocks  \\\n",
       "0  [Phage_T7_tail, unknown_N2]  [unknown_C76, unknown_C234, unknown_C286]   \n",
       "1                           []           [Pectate_lyase_3, Peptidase_S74]   \n",
       "2                [unknown_N82]                             [unknown_C105]   \n",
       "3                           []                             [unknown_C123]   \n",
       "4                [unknown_N48]                                         []   \n",
       "\n",
       "                N_ranges                            C_ranges        unique_ID  \n",
       "0  [(12, 140), (3, 105)]  [(140, 791), (142, 263), (26, 37)]  K39PH122C2_RBP0  \n",
       "1                     []             [(38, 126), (543, 603)]  K39PH122C2_RBP1  \n",
       "2           [(124, 237)]                        [(235, 331)]  K60PH164C1_RBP0  \n",
       "3                     []                         [(66, 115)]  K45PH128C2_RBP0  \n",
       "4              [(0, 82)]                                  []  K45PH128C2_RBP1  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RBPs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e951f19",
   "metadata": {},
   "source": [
    "#### Filter for length & save dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "80db6d89",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_delete = [i for i, protein_seq in enumerate(RBPs['ProteinSeq']) if (len(protein_seq)<200 or len(protein_seq)>1500)]\n",
    "RBPs = RBPs.drop(to_delete)\n",
    "RBPs = RBPs.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "290afa98",
   "metadata": {},
   "outputs": [],
   "source": [
    "RBPs.to_csv(valencia_dir+'/RBPbase'+data_suffix+'.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "512e23c4",
   "metadata": {},
   "source": [
    "#### CD-HIT for 3D structure predictions\n",
    "\n",
    "Here, we cluster the identified RBPs to see how they cluster together. Subsequently, the representative of each of the clusters is subjected to AlphaFold's 3D structure prediction (on high-performance computing infrastructure).\n",
    "\n",
    "Let's keep it simple: all the already known RBPs (with B-helix) that are representatives are white listed by default (and subsequently all the sequences in those clusters), and all other representatives are subjected to a 3D structure prediction for manual check of B-helix presence. If the B-helix is present, we keep all the sequences in the corresponding clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7ddcf600",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdpath = '/Users/dimi/cd-hit-v4.8.1-2019-0228'\n",
    "known_klebsiella_rbps = '/Users/dimi/GoogleDrive/PhD/4_PHAGEHOST_LEARNING/42_DATA/klebsiella_RBP_data/sequences/klebsiella_RBPs.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ab5739e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding known Klebsiella RBPs\n",
    "RBPs = pd.read_csv(valencia_dir+'/RBPbase'+data_suffix+'.csv')\n",
    "combined_fasta = open(valencia_dir+'/RBPs_for_clustering'+data_suffix+'.fasta', 'w')\n",
    "white_list = open(valencia_dir+'/RBPs_white_list'+data_suffix+'.txt', 'w')\n",
    "\n",
    "for record in SeqIO.parse(known_klebsiella_rbps, 'fasta'):\n",
    "    sequence = str(record.seq)\n",
    "    rec_id = record.id\n",
    "    combined_fasta.write('>identified_'+rec_id+'\\n'+sequence+'\\n')\n",
    "    white_list.write('identified_'+rec_id+'\\n')\n",
    "for i, protein in enumerate(RBPs['ProteinSeq']):\n",
    "    unique_id = RBPs['unique_ID'][i]\n",
    "    combined_fasta.write('>'+unique_id+'\\n'+protein+'\\n')\n",
    "combined_fasta.close()\n",
    "white_list.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "522af96c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster with CD-HIT\n",
    "input_file = valencia_dir+'/RBPs_for_clustering'+data_suffix+'.fasta'\n",
    "output_file = valencia_dir+'/RBPs_clusters'+data_suffix\n",
    "cdout, cderr = pu.cdhit_python(cdpath, input_file, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7169b446",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop over clusters: if one in RBPs_with_Bhelix -> white list entire cluster\n",
    "white_list = open(valencia_dir+'/RBPs_white_list'+data_suffix+'.txt')\n",
    "white_list = white_list.read().split('\\n')\n",
    "output_file = valencia_dir+'/RBPs_clusters'+data_suffix\n",
    "clusters = open(output_file+'.clstr')\n",
    "\n",
    "cluster_iter = 0\n",
    "white_check = 0\n",
    "cluster_accessions = []\n",
    "sequences_to_keep = []\n",
    "for line in clusters.readlines():\n",
    "    # new cluster\n",
    "    if line[0] == '>':\n",
    "        # finish old cluster if not first one\n",
    "        if (cluster_iter > 0) and (len(cluster_accessions) >= 1):\n",
    "            if white_check == 1: # an identified RBP in cluster\n",
    "                sequences_to_keep = sequences_to_keep+cluster_accessions\n",
    "        # initiate new cluster\n",
    "        cluster_accessions = []\n",
    "        cluster_iter += 1\n",
    "        white_check = 0\n",
    "    # in a cluster\n",
    "    else:\n",
    "        acc = line.split('>')[1].split('...')[0]\n",
    "        cluster_accessions.append(acc)\n",
    "        if acc in white_list:\n",
    "            white_check = 1\n",
    "        \n",
    "# finish last cluster\n",
    "if len(cluster_accessions) >= 1:\n",
    "    if white_check == 1: # an identified RBP in cluster\n",
    "        sequences_to_keep = sequences_to_keep+cluster_accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d732ec69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "149"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sequences_to_keep)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bf08bdf",
   "metadata": {},
   "source": [
    "## Bacterial genome processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e23d24b",
   "metadata": {},
   "source": [
    "#### Construct dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6436a7eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "klebs_files = listdir(bacterial_genomes_dir)\n",
    "klebs_files.remove('.DS_Store')\n",
    "ERS_list = []; desc_list = []; genome_list = []; strain_list = []\n",
    "\n",
    "for file in klebs_files:\n",
    "    file_dir = bacterial_genomes_dir+'/'+file\n",
    "    genome = ''\n",
    "    descr = ''\n",
    "    for record in SeqIO.parse(file_dir, 'fasta'):\n",
    "        descr = descr + '+' + record.description\n",
    "        genome = genome + str(record.seq)\n",
    "\n",
    "    ERS_list.append(file.split('_')[0])\n",
    "    desc_list.append(descr[1:])\n",
    "    genome_list.append(genome)\n",
    "    strain_list.append(file.split('_')[1].split('.')[0])\n",
    "\n",
    "bacterial_genomes = pd.DataFrame({'GI': ['-']*len(klebs_files),'accession': ERS_list,'description':desc_list,\n",
    "                        'organism': ['Klebsiella pneumoniae']*len(klebs_files),'sequence': genome_list,\n",
    "                        'sequencing_info': ['-']*len(klebs_files),'strain': strain_list,\n",
    "                        'number_of_prophages': ['-']*len(klebs_files)})\n",
    "bacterial_genomes.to_csv(valencia_dir+'/klebsiellaGenomes'+data_suffix+'.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24af34d0",
   "metadata": {},
   "source": [
    "#### Run Kaptive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a972e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_loci(bacterial_genomes, data_dir, project_dir, database_dir, database_name):\n",
    "    \"\"\"\n",
    "    This function uses kaptive_python to loop over all klebsiella_genomes, construct FASTA files for\n",
    "    each of them and identify their loci. Importantly, the file_names are later used to construct embeddings,\n",
    "    so need to be identifiable (accession numbers).\n",
    "        \n",
    "    Input:\n",
    "    - bacterial_genomes: Pandas DataFrame of genomes w/ following column names: 'accession', 'sequence'\n",
    "    - project_directory: the location of kaptive.py (preferrably in the same project folder)\n",
    "    - data_dir: location of the database and sequence file(s) to loop over\n",
    "    - database_name: string of the name of the database (.gbk file)\n",
    "    \"\"\"\n",
    "    kaptive_file_names = []\n",
    "    serotypes = []\n",
    "    pbar = tqdm(total=klebsiella_genomes.shape[0])\n",
    "    \n",
    "    # loop over klebsiella_genomes\n",
    "    for i, genome in enumerate(klebsiella_genomes['sequence']):\n",
    "        #if klebsiella_genomes['number_of_prophages'][i] > 0: # no filter: rows/columns consistent down the road\n",
    "        acc = list(klebsiella_genomes['accession'])[i]\n",
    "\n",
    "        # make FASTA file\n",
    "        file_name = acc+'.fasta'\n",
    "        fasta = open(data_dir+'/'+file_name, 'w')\n",
    "        fasta.write('>'+acc+'\\n'+genome+'\\n')\n",
    "        fasta.close()\n",
    "\n",
    "        # run Kaptive\n",
    "        kaptive_file, _, _ = kaptive_python(project_dir, data_dir, database_name, file_name)\n",
    "        kaptive_file_names.append(kaptive_file)\n",
    "\n",
    "        # process json -> proteins in fasta\n",
    "        results = json.load(open(data_dir+'/kaptive_results.json'))\n",
    "        serotypes.append(results[0]['Best match']['Type'])\n",
    "        protein_results = open(data_dir+'/kaptive_results_proteins_'+acc+'.fasta', 'w')\n",
    "        for gene in results[0]['Locus genes']:\n",
    "            try:\n",
    "                name = gene['Reference']['Product']\n",
    "            except KeyError:\n",
    "                name = 'unknown'\n",
    "            protein = gene['Reference']['Protein sequence']\n",
    "            protein_results.write('>'+name+'\\n'+protein[:-1]+'\\n')\n",
    "        protein_results.close()\n",
    "\n",
    "        # delete FASTA & temp KAPTIVE files\n",
    "        os.remove(data_dir+'/'+file_name)\n",
    "        os.remove(data_dir+'/'+file_name+'.ndb')\n",
    "        os.remove(data_dir+'/'+file_name+'.not')\n",
    "        os.remove(data_dir+'/'+file_name+'.ntf')\n",
    "        os.remove(data_dir+'/'+file_name+'.nto')\n",
    "        os.remove(data_dir+'/kaptive_results.json')\n",
    "\n",
    "        # update progress\n",
    "        pbar.update(1)\n",
    "            \n",
    "    pbar.close()\n",
    "    return kaptive_file_names, serotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3c4f995",
   "metadata": {},
   "outputs": [],
   "source": [
    "database = 'Klebsiella_k_locus_primary_reference.gbk'\n",
    "genomes = pd.read_csv(valencia_dir+'/klebsiellaGenomes'+data_suffix+'.csv')\n",
    "names, serotypes = pu.compute_loci(genomes, project_dir, valencia_dir, database) # results in the same data_dir\n",
    "pd.DataFrame(serotypes, columns=['sero']).to_csv(klebsiella_dir+'/kaptive_serotypes_klebsiella_genomes_031221.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65f232cf",
   "metadata": {},
   "source": [
    "## Interaction matrix construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe3b15b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "IM = pd.read_excel(valencia_dir+'/klebsiella_phage_host_interactions.xlsx', index_col=0, header=0)\n",
    "IM_klebsiella = list(IM.index)\n",
    "IM_phages = [phage.replace(' ', '') for phage in IM.columns]\n",
    "klebsiellaValencia = pd.read_csv(valencia_dir+'/klebsiellaGenomesValencia.csv')\n",
    "RBPbaseValencia = pd.read_csv(valencia_dir+'/RBPbaseValencia.csv')\n",
    "\n",
    "row_names = klebsiellaValencia['accession']\n",
    "column_names = RBPbaseValencia['unique_ID']\n",
    "interactions = np.zeros((len(row_names), len(column_names)))\n",
    "for i, row in enumerate(row_names):\n",
    "    strain = klebsiellaValencia['strain'][i]\n",
    "    row_index = IM_klebsiella.index(strain) # index in IM\n",
    "    for j, col in enumerate(column_names):\n",
    "        phage = RBPbaseValencia['phage_nr'][j]\n",
    "        col_index = IM_phages.index(phage) # index in IM\n",
    "        interactions[i,j] = IM.iloc[row_index, col_index]\n",
    "\n",
    "IMValencia = pd.DataFrame(interactions, index=row_names, columns=column_names)\n",
    "IMValencia.to_csv(valencia_dir+'/interactionsValencia.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('ML1': conda)",
   "language": "python",
   "name": "python397jvsc74a57bd0cdf400e2b2cf645b2ec6a448fcb5b3c1b3d3d5834714944466c9c0370880fa51"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
